### ISSUE TITLE
[$500] [POLISH][Search v2.1] Amount filter - Native - The digit exceeding the max allowed decimal places is briefly inserted into the input

### ISSUE THREAD

**Issue by rayane-d** [Expensify/App#47875]
This is a follow-up issue.
Coming from: https://github.com/Expensify/App/pull/47343#issuecomment-2291467068 and https://github.com/Expensify/App/pull/47343#issuecomment-2298616516

**Bug in iOS and Android native apps:** When entering a number, the digit exceeding the maximum allowed decimal places is briefly inserted into the input field before being deleted. They should not be inserted at all.

https://github.com/user-attachments/assets/5a9be667-53e0-4dc5-bf95-132ad6f09cca

cc @luacmartins @289Adam289 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017b65a6e5c7b77ba2</li>
        <li>Upwork Job ID: 1827006416140957560</li>
        <li>Last Price Increase: 2025-01-08</li>
<li>Automatic offers: </li>
<ul>
<li>brunovjk | Reviewer | 104524279</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @luacmartins</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @luacmartins</details>

**Comment by eyobai** [comment]
Validation Function: Create a function to check if the input exceeds the allowed decimal places.
Event Listener: Use the **input** event to validate each change to the input field.
Immediate Correction: If the input is invalid, immediately revert to the last valid state.

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The digit exceeding the max allowed decimal places is briefly inserted into the input

### What is the root cause of that problem?

This issue occurs because we use onChangeText to validate the amount between lines 37 and line 40. If the amount is not valid, we return and remove the text change in the text input

https://github.com/Expensify/App/blob/0c8455280c738a5db596f34409a0a3177e682e7f/src/components/AmountWithoutCurrencyForm.tsx#L29-L43

### What changes do you think we should make in order to solve the problem?

We should check that the decimal limit has been reached and no more decimals can be entered before the text changes.

```diff
// src/components/AmountWithoutCurrencyForm.tsx#L17
+ const decimals = 2;

// src/components/AmountWithoutCurrencyForm.tsx#L37
- if (!validateAmount(withLeadingZero, 2)) {
+ if (!validateAmount(withLeadingZero, decimals)) {

// src/components/AmountWithoutCurrencyForm.tsx#L45
+     const getDecimalLength = (numStr: string) => {
+        const parts = numStr.split('.');
+        return parts[1] ? parts[1].length : 0;
+    };

// src/components/AmountWithoutCurrencyForm.tsx#L46
    return (
        <TextInput
            value={formattedAmount}
            onChangeText={setNewAmount}
            inputID={inputID}
            name={name}
            label={label}
            defaultValue={defaultValue}
            accessibilityLabel={accessibilityLabel}
            role={role}
            ref={ref}
            keyboardType={CONST.KEYBOARD_TYPE.DECIMAL_PAD}
+           maxLength={getDecimalLength(formattedAmount) === decimals ? formattedAmount.length : undefined}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...rest}
        />
    );
```

<summary>
<details>POC

https://github.com/user-attachments/assets/112bc5dd-67bf-447b-8f6b-e682b7555826

</details>
</summary>

**Comment by brunovjk** [comment]
Thank you for the proposal @huult, but I tested it and it doesn't seem to resolve the issue. The RCA focuses on `maxLength`, but the problem occurs only on native platforms. If you believe your solution fixes this, please share a test branch for review.
For context, I just reproduced the issue on web and native. On web, typing `1111111111111111111111.111111111111` shows `11111111.11` without issue. However, on native (I only tested on iOS so far), the number stays as `1111111111111111111111.111111111111` for a couple of seconds before correcting to `11111111.11`. The issue may require further investigation into why native platforms are behaving differently.

**Comment by AlfredoAlc** [comment]
This most likely is related to a known issue for the `TextInput` component from react-native, specially for iOS platforms. See this issue for reference: https://github.com/facebook/react-native/issues/36494

**Comment by brunovjk** [comment]
Thank you, @AlfredoAlc, for the reference. It aligns with our issue, but I haven’t found a fix yet, only workarounds.
@huult, your proposal appears reasonable as it directly enforces decimal limits. However, despite the straightforward nature of the suggested changes, they haven’t resolved the issue in my tests. Do you have any new findings or a test branch to share? Thank you all!

**Comment by huult** [comment]
@brunovjk , [test branch](https://github.com/huult/App/tree/47875-decimal-input-validation-fix) for my [proposal](https://github.com/Expensify/App/issues/47875#issuecomment-2308487049)

**Comment by brunovjk** [comment]
> @brunovjk , [test branch](https://github.com/huult/App/tree/47875-decimal-input-validation-fix) for my [proposal](https://github.com/Expensify/App/issues/47875#issuecomment-2308487049)

Thank you @huult :D I just tested. Almost there, but the problem is not just in the decimals:

https://github.com/user-attachments/assets/a8bd49c4-8524-41a3-b0f6-c9cd07d4482e

If the video above is no longer available when you watch it, let me know.

**Comment by huult** [comment]
@brunovjk , I've updated the test branch to fix the issue you tested.

**Comment by brunovjk** [comment]
Thank you for your efforts, @huult. Unfortunately, the solution still isn’t working as expected on my end. As we can't attack the root cause directly, we might need a "workaround" to address this effectively. I’m confident we can find one that won’t introduce regressions. Let’s keep exploring options—I’ll revisit this next week, and we can involve an internal engineer if necessary.

https://github.com/user-attachments/assets/756527da-904a-4cf6-8d91-0558d03308e2

**Comment by brunovjk** [comment]
**Update**: Posted a request for help on this issue in [#contributor-plus](https://expensify.slack.com/archives/C02NK2DQWUX/p1725367306529359) to get more insights.

**Comment by isabelastisser** [comment]
Thanks for the update!

**Comment by brunovjk** [comment]
Hey @isabelastisser, @luacmartins, this issue seems to stem from React Native’s handling of inputs on mobile, https://github.com/Expensify/App/issues/47875#issuecomment-2311234992. I suggest we raise compensation to attract more attention, especially for potential fixes. Thoughts on how to proceed? Thanks!

**Comment by luacmartins** [comment]
I think we should keep it as is given it's not a critical issue

**Comment by brunovjk** [comment]
[Bumped on slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1726141986465159) trying to attract more eyes.

**Comment by isabelastisser** [comment]
No proposals yet.

**Comment by isabelastisser** [comment]
Still waiting for proposals.

**Comment by brunovjk** [comment]
Still waiting for proposals.

**Comment by brunovjk** [comment]
Bumped on slack for proposals - https://expensify.slack.com/archives/C01GTK53T8Q/p1727295456196609

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Amount filter - Native - The digit exceeding the max allowed decimal places is briefly inserted into the input

### What is the root cause of that problem?
The root cause is a known issue in controlled react native text input
<img width="822" alt="Screenshot 2024-09-27 at 11 56 31 at night" src="https://github.com/user-attachments/assets/1897e9d9-ae41-45bc-9cfd-0d44d61f6fcc">
When our custom `onChangeText` here will not update the `value` if the decimal value is more than 2 but the react native text input internal state immediately responds and shows the change then it gets back to the unchange `value` prop we are providing
### What changes do you think we should make in order to solve the problem?

As recommended in the react native official page we should use `maxLength` prop to disable inserting digit exceeding max allowed decimal places by calculating the maxLength in `onSelectionChange` event here:
https://github.com/Expensify/App/blob/21f6a048299cbb61d18a5400349deecd62009428/src/components/AmountWithoutCurrencyForm.tsx#L49
 ```
onSelectionChange={(e: NativeSyntheticEvent<TextInputSelectionChangeEventData>) => {
                const selection = e.nativeEvent.selection;
                if (!formattedAmount.includes('.') || selection.end <= formattedAmount.indexOf('.')) {
                    setMaxLength(undefined);
                    return;
                }
                setMaxLength(formattedAmount.split('.')[0].length + 1 + decimal);
            }}
            maxLength={maxLength}
```
We can limit this fix to be applied only for native as the problem occurs only there.
Here is a test [branch](https://github.com/FitseTLT/App/tree/fix-flickering-on-native-for-search-total-field)
POC:

https://github.com/user-attachments/assets/c726b8e2-227e-43d5-957e-2d27e417e45a



### What alternative solutions did you explore? (Optional)

**Comment by brunovjk** [comment]
Thank you very much for the proposal @FitseTLT, it makes sense, RCA seems correct to me. Your solution makes sense, but but after testing your branch does not seem enough to me:

https://github.com/user-attachments/assets/4999887d-c05a-43ad-afaf-f2fb38fd7615

This does not happen on web in my tests. I took a quick look, but I couldn't identify why, do you have any idea? Thanks.

**Comment by FitseTLT** [comment]
@brunovjk I have updated the branch to fix this issue 👍

**Comment by brunovjk** [comment]
Thanks for the update @FitseTLT, [I test it](https://github.com/FitseTLT/App/tree/fix-flickering-on-native-for-search-total-field), and it doesn't seem to have fixed it correctly, were you able to test it without any problems? Maybe I'm doing something wrong.

**Comment by FitseTLT** [comment]
>were you able to test it without any problems? Maybe I'm doing something wrong.

Yep I did. Are you sure you pulled the latest commit?

**Comment by brunovjk** [comment]
Thank you for the update, @FitseTLT! I tested the branch, and it looks better overall. However, there still seems to be a bug: I **When the maximum number is reached**, it is not possible to add decimals. Additionally, the behavior is a bit strange, and I'm having trouble identifying exactly what is happening. 

**Main branch:**

https://github.com/user-attachments/assets/1950dadc-e6f0-4db5-9a0e-c9d8b73b2d2e

**FitseTLT branch:**

https://github.com/user-attachments/assets/4f998a66-87b3-4f17-9465-8ea792cf76cf

However, we still have one detail, when we type a letter in the field, we can still see the flickeringv (also in main). Do you think we can fix that here too? Thank you very much for your effort!

**Comment by isabelastisser** [comment]
@brunovjk, it looks like we're still waiting for proposals, right?

**Comment by brunovjk** [comment]
Exactly. I'll retest next week.

**Comment by FitseTLT** [comment]
@brunovjk both bugs you mentioned are now fixed. Here we go (just don't forget to pull the latest commit on the branch)!! 👍  

https://github.com/user-attachments/assets/c36dea9c-d1ef-474a-bf06-846c14b42287

**Comment by brunovjk** [comment]
Thanks @FitseTLT, I tested the latest changes quickly and it looks promising, I'll review it more carefully during the week and get back to you. Thanks again for the effort.

**Comment by brunovjk** [comment]
I haven't had time to come back here this week yet, tomorrow morning I'll try to test the proposal better.

**Comment by brunovjk** [comment]
Hey @FitseTLT,

Thanks again for your efforts! After closely reviewing the proposed changes, we found a few issues:

- On **web**, no issues before, but now we notice a cursor jump when typing fast or pasting numbers.
- On **native** (iOS), letters and extra digits can still be inserted. We notice cursor jumps when typing fast (e.g., typing "123456" results in "345612"). After further interactions, the app seems to behave as expected. Hard to explain, I made a video, if you have any questions, please let me know.

https://github.com/user-attachments/assets/14984734-5736-4617-84f6-c1d24f0525ae

From my analysis of your changes:
- **Manual cursor calculations** could be causing the jumps when typing fast or pasting numbers.
- **maxLength={1}** seems too restrictive, affecting input behavior.
- **Custom key handling** adds complexity and might conflict with native input behavior, leading to inconsistent results.

But I can't say for sure that these are the problems. Do you want more time to refine this? We can try to ask an internal engineer to review your proposal, and if they like what they see, we can refine and test it properly in the PR. Let me know what you think.

Thanks again!

**Comment by brunovjk** [comment]
Not overdue, still testing solutions

**Comment by brunovjk** [comment]
Same as [above](https://github.com/Expensify/App/issues/47875#issuecomment-2407916309)

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Restricting user input causes flicker.

### What is the root cause of that problem?
This is a bug in RN. Whenever a user makes an input that input is taken into consideration then on the next render (JS side) the text is reverted.

### What changes do you think we should make in order to solve the problem?
Add a new prop to TextInput: `regex`. This prop is used validate user input on native side to respect the specified regex. 

- iOS: Use [textInputShouldChangeText](https://github.com/facebook/react-native/blob/6e120b3b46c73065cd0a86082372e72e8d871946/packages/react-native/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L359) which is internally used by the [shouldChangeCharactersInRange](https://developer.apple.com/documentation/uikit/uitextfielddelegate/1619599-textfield?language=objc) method.
- Android: I have not investigated much here due to build errors but the idea is the same. The initial execution plan is to implement an [InputFilter](https://developer.android.com/reference/android/text/InputFilter) that uses Regex and apply that filter pretty much the same way as done [here](https://github.com/facebook/react-native/blob/0902b0af75ba30ec9d4abeda71769d0685637afa/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L817).

### What alternative solutions did you explore? (Optional)
None

**Comment by s77rt** [comment]
PoC: In RN-Tester, notice the difference between the "no spaces allowed" using the `onChangeText` approach vs the `regex` approach. I have also tested with the regex used in the amount input.


https://github.com/user-attachments/assets/5e348ed7-f819-4e65-9189-4cbab5cad202

**Comment by s77rt** [comment]
@brunovjk Since this is an upstream bug I believe the process is to fix it upstream, for that I'm currently working on RN and I'm planning to raise a PR on RN (or at least a discussion since this is a new feature). For testing you'd have to use RN-Tester. Here is the work-in-progress branch https://github.com/s77rt/react-native/tree/TextInput-regex

**Comment by s77rt** [comment]
Here raised the proposal https://github.com/react-native-community/discussions-and-proposals/pull/826

**Comment by brunovjk** [comment]
Great job @s77rt! We'll keep an eye out then :D Thank you very much, Please let me know if you need anything.
@isabelastisser should we put this issue onhold for @s77rt [proposal on RN community](https://github.com/react-native-community/discussions-and-proposals/pull/826)? Thanks.

**Comment by s77rt** [comment]
I think I'm done with the native implementation, waiting on some feedback on [the proposal](https://github.com/react-native-community/discussions-and-proposals/pull/826). I have also made a [patch version](https://github.com/s77rt/ExpensifyApp/tree/TextInput-regex) for E/App.

| Before | After |
|:------:|:-----:|
|   <video src="https://github.com/user-attachments/assets/1af0f672-feb1-4d52-ac7f-08566417f52a" />     |    <video src="https://github.com/user-attachments/assets/fb1a5a55-b861-471f-ba54-414e0a2c29d6" />  |
|    <video src="https://github.com/user-attachments/assets/18fad594-34ad-4dd7-9349-c01f14f663b4" />    |    <video src="https://github.com/user-attachments/assets/8d258735-c137-4dff-9103-3d8a60c01f6a" />   |

**Comment by brunovjk** [comment]
Great @s77rt! Thanks a lot for the hard work, I'll test it :D

**Comment by brunovjk** [comment]
I’ve tested s77rt's [patch version](https://github.com/s77rt/ExpensifyApp/tree/TextInput-regex) across **Chrome**, the **Desktop** app, **iOS**, and **Android** **native** platforms, and everything is working perfectly:

<details>
<summary>Test results</summary>



https://github.com/user-attachments/assets/c7fc66de-afae-4b7b-9d20-5d19c3c09090


</details>

While the patch performed well in my tests, further discussion and testing by other internal members would be beneficial. You can follow @s77rt's [proposal](https://github.com/Expensify/App/issues/47875#issuecomment-2411866775) along with the related proposal in the RN community: [react-native-community/discussions-and-proposals#826](https://github.com/react-native-community/discussions-and-proposals/pull/826).

🎀👀🎀 C+ reviewed

**Comment by luacmartins** [comment]
Nice! Let's get a PR up for it then

**Comment by brunovjk** [comment]
_This PR was not deployed to staging yet because QA is ongoing_ https://github.com/Expensify/App/pull/51202#issuecomment-2445170521

**Comment by s77rt** [comment]
@brunovjk @luacmartins I had to [revert the PR](https://github.com/Expensify/App/pull/51849) as I found the introduced regression too complex to fix. Main problem that I faced is this one:
- Enter value: `1,900.00`
- Try to remove a digit to make the amount 190
- The intermediate state would be `1,90.00`
- This state is required to pass the regex but it won't because it is not a valid number

=> One regex is not enough
=> Two regex will cause the flickering again

**Comment by brunovjk** [comment]
Thanks for the breakdown, @s77rt. Do you think that are there any alternative approaches or ideas from earlier discussions that we might revisit at this point?

**Comment by s77rt** [comment]
@brunovjk Not really. This bug is in RN and while working on this I had two ideas either use a formatter ([NSNumberFormatter](https://developer.apple.com/documentation/foundation/nsnumberformatter)) which I found more complex to implement or prevent unwanted edits (the regex approach).

The optimal solution would be to support synchronous event handler and then we can do something like running the onChangeText event handler on the native side and use its return value as the final input value

**Comment by brunovjk** [comment]
We are [reverting the PR](https://github.com/Expensify/App/pull/51849) for [this](https://github.com/Expensify/App/issues/51800#issuecomment-2451740709) reason

**Comment by brunovjk** [comment]
@s77rt do you have any idea how we can solve this issue now? I couldn't find a reliable enough workaround. Thank you :D

**Comment by s77rt** [comment]
@brunovjk Unfortunately we may just want to close this issue for now. React Native events callback are async and this delay(flickering) is something to expect. The bug is not new as it occurs in all places where we format user input e.g. room name. A solution to prevent this flickering is to not format the input and instead show an error that the number is not valid. But I don't think this is worth it either.


https://github.com/user-attachments/assets/d7de883a-542b-4a73-b521-3eac3f1fd531

**Comment by brunovjk** [comment]
Not thrilled about closing this without a fix, but I get the challenge here. I tested some of the solutions proposed by others, but none seemed reliable enough in my trials. Thanks for the insights, @s77rt. @luacmartins and @isabelastisser, any thoughts?

**Comment by luacmartins** [comment]
I understand the challenge in fixing this bug, given it's nature, but I'd keep this issue open in case we get any other proposals.

**Comment by brunovjk** [comment]
Update: Posted on Slack for proposals https://expensify.slack.com/archives/C01GTK53T8Q/p1730980932936309

**Comment by brunovjk** [comment]
Update: Still looking for solutions

**Comment by isabelastisser** [comment]
No updates yet!

**Comment by luacmartins** [comment]
I'm gonna bump the price since we're not getting proposals

**Comment by brunovjk** [comment]
**Update:** Initially, we identified the issue as a known problem with the `TextInput` component in React Native (https://github.com/facebook/react-native/issues/36494). To date, we have received three proposals:

--------- 
### @huult's [proposal](https://github.com/Expensify/App/issues/47875#issuecomment-2308487049):
<details>
<summary>details</summary>

### Proposal:
Suggested enforcing decimal limits dynamically by validating changes in `onChangeText`, adding a utility to check decimal length, and restricting `maxLength` for the input.

### Outcome:
Extra digits were briefly visible before being corrected. Updates to the proposal improved handling, but didn’t fully resolve the issue, especially for pasted values and multiple decimal points.

</details>

### @FitseTLT's [proposal](https://github.com/Expensify/App/issues/47875#issuecomment-2380315791)
<details>
<summary>details</summary>

### Proposal:
The solution involves calculating the `maxLength` dynamically based on the position of the decimal point, ensuring that no extra digits are inserted after the allowed decimal places.

### Outcome:
Extra digits and letters could still be inserted. Cursor jumps occurred when typing quickly (e.g., "123456" became "345612").

</details>

### @s77rt's [proposal](https://github.com/Expensify/App/issues/47875#issuecomment-2411866775)
<details>
<summary>details</summary>

### Proposal:
This approach proposes adding a new `regex` prop to the `TextInput` component, which would validate user input on the native side using a regular expression.

### Outcome:
- **Patch Testing**: The patch was tested across various platforms, including Chrome, Desktop, iOS, and Android. It worked reliably on the native environments.
- **Challenges**: Despite successful tests, implementing two regex validations or synchronous event handling remains unresolved. Leading to some regressions: https://github.com/Expensify/App/issues/51800 and https://github.com/Expensify/App/issues/51802. Solutions like using a formatter (`NSNumberFormatter`) were deemed too complex and could potentially introduce issues like flickering.

For reference, a proposal was raised in the React Native community: [react-native-community/discussions-and-proposals#826](https://github.com/react-native-community/discussions-and-proposals/issues/826).

</details>
--------- 

If you have different outcomes, please share your findings. Thank you all.

**Comment by kirillzyusko** [comment]
I think the correct solution would be to use mask components, such as `react-native-text-input-mask`. A mask could look like `…{.}[99]` In this case formatting will happen on a native side (extra symbols will not be typed at all) and you don't need to create a complex solutions in JS 🤷‍♂️ 

@brunovjk have you ever considered this option? If yes, then why it didn't work out for you?

**Comment by brunovjk** [comment]
Thank you for the suggestion! No, we haven't explored input-mask yet, but it seems like a good fit. Would you mind creating a proposal with a proof-of-concept or a branch so we can present it to the team for evaluation? 🚀

**Comment by kirillzyusko** [comment]
> Would you mind creating a proposal with a proof-of-concept or a branch so we can present it to the team for evaluation? 🚀

Yes, sure, I can try 🙌 

I'm currently busy with other tasks in Expensify (that have higher priority), but I hope I can get some time early next week. Would you mind to assign me on this task? I don't think I can provide daily updates (just because busy with other tasks), but I think I should be able to present a working PoC next week 👀

**Comment by brunovjk** [comment]
Nice! Let's try to create a proposal next week :D I will try make some tests here.

**Comment by brunovjk** [comment]
I can still reproduce this issue: 

https://github.com/user-attachments/assets/e713c754-f228-4c93-a6d2-38d300e31586

Looking forward to proposals/solutions.

**Comment by brunovjk** [comment]
Same as above

**Comment by kirillzyusko** [comment]
@brunovjk I created a PoC here: https://github.com/Expensify/App/pull/53695

Feel free to play with that changes and if you have any questions I'll be happy to answer them 😊 

It's worth to note, that I didn't integrate a new type of input (masked input) into existing codebase. I simply showed how masked input can work - so styling of input is broken and most likely form submitting mechanism is broken as well, it's just a PoC to show, that masked input will not allow you to enter more symbols if mask template has been filled properly.

If we decided to go with this approach (and I really hope we will do that) then I'll properly integrate the new input into existing codebase so that we can use it preserving all styles and all generic components that we already use 🤞

**Comment by brunovjk** [comment]
Great @kirillzyusko! I'll review it calmly on Monday 🤞 and let you know, but in any case we eventually need a proposal, following the [template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md), to present to the internal team, only they, after a proposal is approved, can assign you for the task. But I'm already very grateful for your effort.

**Comment by brunovjk** [comment]
Testing today

**Comment by brunovjk** [comment]
Thanks for providing the POC, @kirillzyusko. I tested the implementation, but unfortunately, it doesn't seem to work as expected on my side. I don’t think I missed anything, but there’s a chance something went wrong during my setup. Could you share how your testing went? Thank you! 

https://github.com/user-attachments/assets/32292e52-f852-4ae6-9df4-ec1a03a93209

**Comment by kirillzyusko** [comment]
@brunovjk try to set a dark theme and you will see the white input below (in the light theme the white input is barely visible).

You are testing old implementation. On this page I re-worked only one input (the second one) and kept first one to take it as a reference in terms of functionality formatting rules.

Would you mind to do testing again? I'll share a video tomorrow 🙃

**Comment by brunovjk** [comment]
Thanks for clarifying, testing in a moment

**Comment by kirillzyusko** [comment]
> do you believe these bugs were expected and we will fix them later?

Well, I believe I can fix some of them straight away (such as `,` and `.` separator or mask adjustments) and pasting probably would require a little bit more time but also should be fixable 🤞 

My main concern is that we would need to add a new library... Does it make sense to open an issue about adding a new library and see whether Expensify team is happy to have it in `package.json` or not?

**Comment by brunovjk** [comment]
Great question @kirillzyusko, I'm going to ask you a favor, write a proposal with all the information you think is important, along with these questions, we will forward it to an internal member of the Expensify team, they knew exactly what to do next. How does it sound?

**Comment by kirillzyusko** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The digit exceeding the max allowed decimal places is briefly typed into the input

### What is the root cause of that problem?

The problem in current implementation is that the communication between JS/Native thread is asynchronous. So when we type a value we immediately apply update on a native thread and show a newly typed digit. But formatting happens on JS, thus we always have a delay of 16ms and we can observer how digits are flickering.

### What changes do you think we should make in order to solve the problem?

We should use a solution, where formatting happens synchronously on a native thread.

In `react-native` this problem is solved long time ago using [react-native-text-input-mask](https://www.npmjs.com/package/react-native-text-input-mask). This library relies on native libraries developed by [RedMadRobot](https://github.com/RedMadRobot) for [android](https://github.com/RedMadRobot/input-mask-android) and [ios](https://github.com/RedMadRobot/input-mask-ios) (`react-native-text-input-mask` acts as a simple wrapper).

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?

It can be hard to automate this thing, just because only humans can notice the flickering.

### What alternative solutions did you explore? (Optional)

`react-native-text-input-mask` can not be integrated into Expensify project just because it's abandoned project and it wasn't updated for a quite long time (this lib doesn't support Fabric architecture yet).

As an alternative we can use a fork called [react-native-advanced-input-mask](https://github.com/IvanIhnatsiuk/react-native-advanced-input-mask). Advantages of this fork:
- constantly gets updates;
- support both fabric and paper architectures;
- new features are in development (for example combining mask-placeholder with the text that you typed).

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

@brunovjk please, review this proposal and in case if anything is missing - let me know 😊

**Comment by brunovjk** [comment]
I asked on [Slack](https://expensify.slack.com/archives/D08570DK8E5/p1734008492733969) and luacmartins will look into it when they have a moment.

**Comment by kirillzyusko** [comment]
@brunovjk please, keep me updated here as well 🙌 It looks like I don't have an access to this channel/conversation 😔

**Comment by brunovjk** [comment]
Of course, we are still waiting for @luacmartins's thoughts.

**Comment by luacmartins** [comment]
So it seems like the suggestion is to go with the alternate solution, given the main library is abandoned and doesn't support fabric. I'll raise the library request in Slack and we'll see what the team decides

**Comment by luacmartins** [comment]
Created an issue for the lib request [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1734533800254579) and started a [discussion in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1734533800254579)

**Comment by isabelastisser** [comment]
### FYI, I will be OOO from Dec 23 to Jan 6, so please reassign the issue to another team member for urgency, IF needed.

**Comment by brunovjk** [comment]
Still under [discussion on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1734533800254579)

**Comment by brunovjk** [comment]
Still under [discussion on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1734533800254579)

**Comment by brunovjk** [comment]
> Still under [discussion on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1734533800254579)

Same

**Comment by brunovjk** [comment]
Not overdue, still under [discussion on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1734533800254579)

**Comment by brunovjk** [comment]
Not overdue

**Comment by kirillzyusko** [comment]
@MelvinBot not overdue. I'm working on a fix in https://github.com/Expensify/App/pull/53695 Last remaining thing is to make it able to paste values with commas and automatically transform it to dot format. Currently working on that!

**Comment by brunovjk** [comment]
Not overdue :D

**Comment by kirillzyusko** [comment]
> I'm working on a fix in https://github.com/Expensify/App/pull/53695 Last remaining thing is to make it able to paste values with commas and automatically transform it to dot format. Currently working on that!

Fixed the problem with pasting. PR https://github.com/Expensify/App/pull/53695 is ready for review 👀

**Comment by brunovjk** [comment]
**Update**: We are still working on the PR.

**Comment by brunovjk** [comment]
Do not proceed with payment yet, we are working on this additional [PR](https://github.com/Expensify/App/pull/56462).

**Comment by brunovjk** [comment]
> we are working on this additional [PR](https://github.com/Expensify/App/pull/56462).

**Update**: Same as above

**Comment by brunovjk** [comment]
PR is on staging

**Comment by luacmartins** [comment]
not overdue, PR in staging

**Comment by luacmartins** [comment]
PR deployed to production

**Comment by luacmartins** [comment]
Payment is being handled in https://github.com/Expensify/App/issues/56330, so I'm closing this issue
